kimchi/src/circuits/expr.rs

```diff
@@ -3149,3 +3149,246 @@ pub mod test {
+            self,
+            evals: &crate::proof::ProofEvaluations<crate::proof::PointEvaluations<F>>,
+        ) -> Option<crate::proof::PointEvaluations<F>> {
+            match self {
+                ColumnExtended::Inner(s) => s.evaluate(evals),
+                ///TODO: a change in evals will be required so that it can also provide access to the extended columns
+                ColumnExtended::Extended(e) => todo!(),
+            }
+        }
+
+        fn make_witness(i: usize) -> Self {
+            Self::Inner(S::make_witness(i))
+        }
+
+        fn make_coefficient(i: usize) -> Self {
+            Self::Inner(S::make_coefficient(i))
+        }
+    }
+    ///records expressions that have been extracted into an extra column
+    struct ExprRecorder<C, Col>
+    where
+        C: Hash + Eq,
+        Col: ColTrait + Hash + Eq,
+    {
+        recorded_exprs: HashMap<Expr<C, ColumnExtended<Col>>, usize>,
+        next: usize,
+    }
+
+    impl<C: Hash + Eq, Col: ColTrait + Hash + Eq> ExprRecorder<C, Col> {
+        fn new() -> Self {
+            Self {
+                recorded_exprs: Default::default(),
+                next: 0,
+            }
+        }
+        fn get_id(&mut self, e: Expr<C, ColumnExtended<Col>>) -> usize {
+            *self.recorded_exprs.entry(e).or_insert_with(|| {
+                let id = self.next;
+                self.next += 1;
+                id
+            })
+        }
+        fn into_constraints(self) -> Vec<Expr<C, ColumnExtended<Col>>> {
+            let ExprRecorder { recorded_exprs, .. } = self;
+            let mut new_constraints = BTreeMap::new();
+            for (exp, id) in recorded_exprs.into_iter() {
+                let left = Box::new(extended(id));
+                let constraint =
+                    Expr::<C, ColumnExtended<Col>>::BinOp(Op2::Sub, left, Box::new(exp));
+                new_constraints.insert(id, constraint);
+            }
+            new_constraints.into_values().collect()
+        }
+    }
+
+    fn extended<C, Col: ColTrait>(id: usize) -> Expr<C, ColumnExtended<Col>> {
+        Expr::Cell(Variable {
+            col: ColumnExtended::Extended(id),
+            row: CurrOrNext::Curr,
+        })
+    }
+
+    impl<C, Col> Expr<C, Col>
+    where
+        C: Zero + One + Neg<Output = C> + PartialEq + Clone + Hash + Eq + Debug,
+        Col: ColTrait + Hash + Eq,
+    {
+        /// returns self transformed into a degree 2 or less constraint and a set of additional constraints,
+        /// each for one of the new columns
+        pub fn reduce_degree_to_2(
+            self,
+        ) -> (
+            Expr<C, ColumnExtended<Col>>,
+            Vec<Expr<C, ColumnExtended<Col>>>,
+        ) {
+            let mut rec = ExprRecorder::new();
+            let exp = self.reduce_degree_to_2_rec(&mut rec);
+            let new_constraints = rec.into_constraints();
+            (exp, new_constraints)
+        }
+        fn pow_to_mul(self, p: u64) -> Self {
```

Comment:
Maybe generically would look better, like do a bit decomposition and then in the loop multiply the accumulator by the power of `e` if necessary. I think it also doesn't create more expressions than necessary.